quiche/src/tests.rs

```diff
@@ -1444,34 +1444,249 @@ fn flow_control_update(
+    assert_eq!(pipe.server.rx_data, 30);
+
+    // client sends frame that partially closes the gap
+    // now we have [0..8] and [10..15]
+    // verify flow control. Should be no change.
+    send_frame_helper(&mut pipe, RangeBuf::from(&[1; 3], 5, false)).unwrap();
+    assert_eq!(pipe.server.rx_data, 30);
+    assert_eq!(pipe.server.flow_control.consumed(), 15);
+
+    // client sends partially overlapping data and partially new data
+    // now we have [0..8] and [10..20]
+    // verify flow control. we should account for an additional 5 bytes
+    send_frame_helper(&mut pipe, RangeBuf::from(&[1; 10], 10, false)).unwrap();
+    assert_eq!(pipe.server.rx_data, 35);
+    assert_eq!(pipe.server.flow_control.consumed(), 20);
+
+    // client sends a fin, but again with a gap
+    // this should add another 5 bytes to flow control
+    send_frame_helper(&mut pipe, RangeBuf::from(&[0; 0], 25, true)).unwrap();
+    assert_eq!(pipe.server.rx_data, 40);
+    assert_eq!(pipe.server.flow_control.consumed(), 25);
+}
+
+#[rstest]
+/// Tests that flow control is properly updated when a stream receives a RESET
+fn flow_control_reset_stream(
+    #[values("cubic", "bbr2", "bbr2_gcongestion")] cc_algorithm_name: &str,
+    #[values("reset", "fin")] inconsistent_final_size_frame: &str,
+) {
+    let mut buf = [0; 65536];
+
+    let mut config = Config::new(PROTOCOL_VERSION).unwrap();
+    assert_eq!(config.set_cc_algorithm_name(cc_algorithm_name), Ok(()));
+    config
+        .load_cert_chain_from_pem_file("examples/cert.crt")
+        .unwrap();
+    config
+        .load_priv_key_from_pem_file("examples/cert.key")
+        .unwrap();
+    config
+        .set_application_protos(&[b"proto1", b"proto2"])
+        .unwrap();
+    // Set large initial max_data so we don't have to deal with MAX_DATA
+    // or STREAM_MAX_DATA frames
+    config.set_initial_max_data(15_000);
+    config.set_initial_max_stream_data_bidi_local(15_000);
+    config.set_initial_max_stream_data_bidi_remote(15_000);
+    config.set_initial_max_stream_data_uni(10);
+    config.set_initial_max_streams_bidi(3);
+    config.set_initial_max_streams_uni(3);
+    config.set_max_idle_timeout(180_000);
+    config.verify_peer(false);
+    config.set_ack_delay_exponent(8);
+
+    let mut pipe = test_utils::Pipe::with_config(&mut config).unwrap();
+    assert_eq!(pipe.handshake(), Ok(()));
+
+    // Client opens a stream and sends some data.
+    assert_eq!(pipe.client.stream_send(0, b"aaaaa", false), Ok(5));
+    assert_eq!(pipe.advance(), Ok(()));
+
+    // Server receives data, without reading it.
     let mut r = pipe.server.readable();
+    assert_eq!(r.next(), Some(0));
     assert_eq!(r.next(), None);
 
-    // Flush connection.
-    assert_eq!(pipe.advance(), Ok(()));
+    // check flow control accounting
+    assert_eq!(pipe.server.rx_data, 5);
+    assert_eq!(pipe.server.flow_control.consumed(), 0);
+
+    // Helper function that sends STREAM frames to the server on stream 4
```

Comment:
The comment is a lie! :cake: 

There is already a send_pkt_to_server helper, so I'm wondering if this really does anything different and we can cut down the code a bit. 

If it is different for some reason, we should at least cut out the duplicate helper in flow_control_drain via extraction/parameterization